File: src/test/java/dev/daae/time/config/ExceptionHandlerTest.java
package dev.daae.time.config;

import static org.mockito.Mockito.when;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.httpBasic;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import dev.daae.time.IntegrationTest;
import dev.daae.time.repositories.SessionRepository;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.web.servlet.MockMvc;

@AutoConfigureMockMvc
class ExceptionHandlerTest extends IntegrationTest {

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private SessionRepository sessionRepository;

    @Test
    void statusEndpointReturns500WhenLogRepositoryThrowsException() throws Exception {
        when(sessionRepository.findFirstByOrderByStartDesc()).thenThrow(new RuntimeException());
        this.mockMvc.perform(get("/status").with(httpBasic("username", "password")))
            .andExpect(status().isInternalServerError())
            .andExpect(jsonPath("$.error").value("Internal server error"));
    }

    @Test
    void statusEndpointReturns500WhenLogRepositoryThrowsNullPointerException() throws Exception {
        when(sessionRepository.findFirstByOrderByStartDesc()).thenThrow(new NullPointerException());
        this.mockMvc.perform(get("/status").with(httpBasic("username", "password")))
            .andExpect(status().isInternalServerError())
            .andExpect(jsonPath("$.error").value("Internal server error"));
    }
}

File: src/test/kotlin/dev/daae/time/config/ExceptionHandlerTest.kt
package dev.daae.time.config

import dev.daae.time.IntegrationTest
import dev.daae.time.models.Session
import dev.daae.time.repositories.SessionRepository
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import java.lang.Exception
import java.lang.NullPointerException
import java.lang.RuntimeException

@AutoConfigureMockMvc
internal class ExceptionHandlerTest() : IntegrationTest() {
    @Autowired
    private val mockMvc: MockMvc? = null

    @MockitoBean
    private val sessionRepository: SessionRepository? = null

    @Test
    @Throws(Exception::class)
    fun statusEndpointReturns500WhenLogRepositoryThrowsException() {
        Mockito.`when`<Session?>(sessionRepository!!.findFirstByOrderByStartDesc()).thenThrow(RuntimeException())
        this.mockMvc!!.perform(
            MockMvcRequestBuilders.get("/status")
                .with(SecurityMockMvcRequestPostProcessors.httpBasic("username", "password")),
        )
            .andExpect(MockMvcResultMatchers.status().isInternalServerError())
            .andExpect(MockMvcResultMatchers.jsonPath("$.error").value("Internal server error"))
    }

    @Test
    @Throws(Exception::class)
    fun statusEndpointReturns500WhenLogRepositoryThrowsNullPointerException() {
        Mockito.`when`<Session?>(sessionRepository!!.findFirstByOrderByStartDesc()).thenThrow(NullPointerException())
        this.mockMvc!!.perform(
            MockMvcRequestBuilders.get("/status")
                .with(SecurityMockMvcRequestPostProcessors.httpBasic("username", "password")),
        )
            .andExpect(MockMvcResultMatchers.status().isInternalServerError())
            .andExpect(MockMvcResultMatchers.jsonPath("$.error").value("Internal server error"))
    }
}
